Add ChargeIncrementModelHandler

j-wags opened the pull request:
- [x] Add support for [ChargeIncrementModel tag from the SMIRNOFF spec docs](https://open-forcefield-toolkit.readthedocs.io/en/latest/smirnoff.html#chargeincrementmodel-small-molecule-and-fragment-charges)
- [x] Adds `raise_exception_types` kwarg to `ToolkitRegistry.call`, enabling it to optionally intercept or pass through certain exceptions. This is needed in this PR, as many charge-calculation methods will be accessed through the `assign_partial_charges(partial_charge_method)` method in different toolkits, and each toolkit needs to be able signal to the `ToolkitRegistry` that a charge method is unrecognized but that other toolkits might support it. 
- [x] Implement `AmberToolsToolkitWrapper.assign_partial_charges(mol, partial_charge_method)` for supported methods
- [x] Implement `OpenEyeToolkitWrapper.assign_partial_charges(mol, partial_charge_method)` for supported methods
- [x] Implement `BuiltInToolkitWrapper` class, with `assign_partial_charges(mol, partial_charge_method)` supporting `zeroes` and `formal_to_partial_charge`
- [x] Implement #208 
- [x] Resolve #348 
- [x] Resolve #465 
- [x] Resolve #492 
- [x] Fix #532
- [x] Puts `temporary_cd` back because it turns out we actually do use it (#595) 
- [x] Add [tests](https://github.com/openforcefield/openforcefield/tree/master/openforcefield/tests)
- [x] Update docstrings/[documentation](https://github.com/openforcefield/openforcefield/tree/master/docs), if applicable
- [x] Update [changelog](https://github.com/openforcefield/openforcefield/blob/master/docs/releasehistory.rst)

jchodera commented on The-SMIRNOFF-force-field-format.md:
```diff
@@ -233,19 +233,19 @@ Here is an example:
 * The `number_of_conformers` attribute (default: `"10"`) is used to specify how many conformers will be generated for the molecule (or capped fragment) prior to charging.
 * The `quantum_chemical_method` attribute (default: `"AM1"`) is used to specify the quantum chemical method applied to the molecule or capped fragment.
-* The `partial_charge_method` attribute (default: `"CM2"`) is used to specify how uncorrected partial charges are to be generated from the quantum chemical wavefunction. Later additions will add restrained electrostatic potential fitting (RESP) capabilities.   
+* The `partial_charge_method` attribute (default: `"Mulliken"`) is used to specify how uncorrected partial charges are to be generated from the quantum chemical wavefunction. Later additions will add restrained electrostatic potential fitting (RESP) capabilities.   
```
In our last discussion, I think we wanted to combine `quantum_chemical_method` and `partial_charge_method` to instead have a general two-stage procedure:
* Stage 1: Produce initial charges via some QM or ML method: AM1-Mulliken, graph convolutional network, VCharge, etc.
* Stage 2: Correct the initial charges with BCCs

That will require more significant revision of the proposed ChargeIncrementModel spec.

j-wags replied:
Ok, I thought we had been trending toward this decision, but I couldn't find it in writing anywhere. I'm going to have `partial_charge_method` be the winner here (so, for example, `partial_charge_method="AM1-Mulliken"`). FOr the ToolkitRegistry, each toolkit will have a `supports_charge_method(method)` function which takes a `method` string and returns a bool indicating whether it can calculate charges using that method.

jchodera replied:
The partial charge method may itself have other parameters, so it might be useful to split this into two separate tags for the different stages.

When we had discussed this, I think we discussed generalizing it into perhaps two tags: an initial charging tag (that might take optional parameters) and a BondChargeIncrement tag (that describes how to modify charges). 

Happy to chat further if you want to discuss ideas for making this flexible and simple to implement.

jchodera commented on The-SMIRNOFF-force-field-format.md:
````diff
@@ -231,21 +231,20 @@ In keeping with the AMBER force field philosophy, especially as implemented in s
 Here is an example:
 ```XML
-<ChargeIncrementModel version="0.3" number_of_conformers="10" quantum_chemical_method="AM1" partial_charge_method="CM2">
+<ChargeIncrementModel version="0.3" number_of_conformers="1" partial_charge_method="AM1-Mulliken">
````
I still think we may want to separate this into two tags, since even AM1-Mulliken methods may expose options such as
* Whether to use multiple conformers (e.g. AM1 vs ELF AM1)
* How many conformers to generate (e.g. ELF vs ELF10)
* What tolerance to use for minimization
We may not want to cram tons of additional attributes into this single tag to keep qualifying the 

Also, the `number_of_conformers` attribute will no longer be relevant when we move to graph convolutional networks. Instead, it may be a model name.

j-wags replied:
Just to record what we discussed earlier this week -- My thought is that there are so many different ways this tag could go, that I'm just going to make it a string `partial_charge_method` and an int `number_of_conformers` in the initial version. This will almost certainly change in the future, but we can modify the spec once we have a clearer idea of the scope.

j-wags commented:
I took some time to think about this today, and wanted to update folks on the status of this PR, and ask for feedback on a few topics. 

### Design decisions I'm pretty certain about

There is no clear "happy medium" for the accepted keywords in our charge method spec. For the initial release, I'm going with `partial_charge_method` and `number_of_conformers`, which I hope will last us some time. But if we do a good job of enabling experimentation, people will ask for more keywords like `conformer_rms_cutoff`, `conformer_charge_averaging_method`, `basis_set`, `qm_method`, etc... and we'll either have to spend a ton of effort being the gatekeeper for spec changes, end up allowing a huge number of keywords (requiring complex validation logic), or if we're permissive with changes, community PRs will make the section spec version change monthly. Until there's more clarity on which direction this will go, I'm pretty certain that `partial_charge_method` should simply be a string that is accepted by a ToolkitWrapper's `compute_partial_charges` method. 

### Current state of this functionality in this branch

* `partial_charge_method`is a string that users can populate with supported keywords (currently `AM1-Mulliken`, `AM1BCC`, `AM1BCCELF`, `Gasteiger`, `formal_charge`, `zeroes`, etc...), and folks interested in experimenting can add newly-supported keywords as they see fit. 
* `number of conformers` is an int that users can set, but only one charge method (AM1BCCELF) can actually do anything with multiple conformers (and it only supports "10"). All others simply require `number_of_conformers` to be 1 (geometry-dependent methods) or 0 (`Gasteiger`, `formal_charges`, `zeroes`)
* In the future, we should switch to define charge methods with a `quantum_chemical_method` field that accepts dicts representing QCSchema-defined objects (or some subset thereof) to define how we get the molecule wavefunction, and then another set of keywords to define what to do with that wavefunction to get partial charges. 
* The user has basically no control over the conformers that are used for charge gen other than saying how many should be generated. Even if they provide conformers, we re-generate them in the handler (since we don't know where their conf came from, and prefer to provide "deterministic" charges). 
    * If a user really wants to provide their own conf(s) for charge gen, they can do so by running `molecule.compute_partial_charges` with their desired confs _before_ `create_openmm_system`, and then list that molecule in the `charge_from_molecules` kwarg

### Open questions for all:
* What is proper behavior when a chargeincrement is applied on top of symmetric groups? For example, should chargeincrements applied to a methyl with the SMARTS `[C:1]([H:2])([H:3])([H:4])` apply once,  three times, or six times? If once, how could we programmatically deduplicate this match? (`ValenceDict` won't save us here, since the order of the atoms in the tuple IS important)
* Should `ChargeIncrement` and `ChargeMethod` be separate headers, with `ChargeIncrement` being allowed to apply to many charge methods? (More thoughts on this in the next section)


### Questions for @proteneer:
@proteneer will be one of the major users of this, and I'm not sure that the draft implementation can cover what he aims to do -- 
* Our charge increments are more general than BCCs - They can include more than two atoms, so they define `n` charge increments for transferring charge between `n` atoms. I recall that @proteneer was optimistic about using _bond_ increments since they could be defined with a single increment value, which could be varied while preserving the desired net charge on the molecule. @proteneer, could you let me know if you can deal with `n` charge increments for `n` atoms, or if we should rethink this functionality for your uses? 
* The current implementation can't apply increments on top of charges from other sources in the FF (I think @proteneer had talked about applying these on top of `LibraryCharges` or `charge_from_molecules`). If this is still needed, we could either 
    * (I'd prefer this) allow defining `apply_to` (a list of strings) in the header section of `ChargeIncrements`, which would be a list of charge method tags that charge increments should be applied on top of, and switch out semiempirical calcs to new header "ChargeMethod" 
    * (I wouldn't prefer this) allow defining `ChargeIncrement` blocks inside of other SMIRNOFF data headers

j-wags commented:
@proteneer and I just talked, and came to the following conclusions:

1) We should use the term "Atom Charge Correction (ACC)" when talking about what we're doing in the `ChargeIncrementModel` section. SMARTS-based charge transfer is not restricted to bonds, so "BCC" is inaccurate. The crux of my question above is that ACCs require an equal number of tagged atoms and charge increments, whereas BCCs require N-1 charge increments. One advantage of BCCs is that you can throw them into naive optimization schemes, and always have the net charge on the molecule be 0. * @proteneer said that there are already spots in his optimization framework that have interdependent parameters, so adding something like `chargeincrement2 = - chargeincrement1` is not overly cumbersome to do.

2) @proteneer pointed out that, if we wanted to apply `ChargeIncrement`s to charges from `LibraryCharges` or `charge_from_molecules`, then we could have just changed those values in the original definition (to have the adjustments already applied). So, there's not currently a need for him to apply charge increments to charges from other sources.

So, in the current implementation:
* `ChargeIncrementModel` will remain as a single section in SMIRNOFF (it won't be split into `ChargeModel` and `ChargeIncrements`)
* Charge increments will only apply to the charge model in the `ChargeIncrementModel` section
* We will implement "ACC"s, requiring one charge increment for each tagged atom. We will not implementing BCCs, where you can provide N-1 charge increments for N atoms.

Still unresolved is the symmetry question, but if nobody has strong feelings about it, I'll just sit down, do something reasonable, and mention that behavior in the spec and release notes.

jchodera commented:
Following that logic, wouldn't these be "group charge increments" since the increment for one or a connected group of atoms?

j-wags commented:
> Following that logic, wouldn't these be "group charge increments" since the increment for one or a connected group of atoms?

That'd also be fine, though I'm partial to "ACC" since first-time users will have an easier time of understanding the concept by relation to "BCC"s. I'm slightly in favor of ACC for this reason, but am happy to change if you strongly prefer "group charge increments".

proteneer commented:
+1 on ACCs, not sure if group refers to atom or bonds

On Fri, Jan 10, 2020 at 3:47 PM Jeff Wagner <notifications@github.com>
wrote:

> Following that logic, wouldn't these be "group charge increments" since
> the increment for one or a connected group of atoms?
>
> That'd also be fine, though I'm partial to "ACC" since first-time users
> will have an easier time of understanding the concept by relation to
> "BCC"s. I'm slightly in favor of ACC for this reason, but am happy to
> change if you strongly prefer "group charge increments".
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/openforcefield/openforcefield/pull/471?email_source=notifications&email_token=AARM2FB75T54Z4JA34QQBYTQ5DNGHA5CNFSM4JVT37FKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIVFFUY#issuecomment-573199059>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARM2FB67RPVFZ6V3IP2L73Q5DNGHANCNFSM4JVT37FA>
> .
>
-- 
Yutong Zhao

j-wags commented on openforcefield/tests/test_toolkits.py:
```diff
@@ -657,77 +659,221 @@ def test_generate_conformers(self):
             charge_sum += pc
-        assert charge_sum < 0.001 * unit.elementary_charge
+            abs_charge_sum += abs(pc)
+        assert abs(charge_sum) < 0.005 * unit.elementary_charge
```
I had to increase the threshold because AmberTools only records to 3 digits of precision in charges.

j-wags commented on openforcefield/typing/engines/smirnoff/forcefield.py:
```diff
@@ -885,8 +885,10 @@ def _load_smirnoff_data(self, smirnoff_data, allow_cosmetic_attributes=False):
-                parameter_list_dict = {parameter_list_tagname: parameter_list_dict}
+
+                # If the parameter list isn't empty, it must be transferred into its own tag.
+                if parameter_list_dict != {}:
```
This is necessary for deserializing SMIRNOFF force field sections which may or may not have any smirks-based elements (like an empty `ChargeIncrementModel` tag)

j-wags commented on openforcefield/utils/toolkits.py:
```diff
@@ -142,8 +157,6 @@ class ToolkitWrapper:
     _is_available = None  # True if toolkit is available
     _toolkit_name = None  # Name of the toolkit
     _toolkit_installation_instructions = None  # Installation instructions for the toolkit
-    _toolkit_file_read_formats = None  # The file types that this toolkit can read
-    _toolkit_file_write_formats = None  # The file types that this toolkit can write
```
There was ambiguity as to whether these were CLASS or INSTANCE methods. RDKit did something silly where this is actually a dict like `{'sdf':rdkit.Chem.SDFWriterClass}`, so even so much as putting the class definition through the interpreter would raise an exception if rdkit wasn't available. Since there are a few cases where this could benefit from being modified, I figure it's best to switch it over to an instance attribute (defined in `__init__`)

j-wags commented on openforcefield/utils/utils.py:
```diff
@@ -383,17 +405,25 @@ def check_units_are_compatible(object_name, object, unit_to_check, context=None)
     IncompatibleUnitError
     """
     from simtk import unit
+
+    # If context is not provided, explicitly make it a blank string
+    if context is None:
+        context = ''
+    # Otherwise add a space after the end of it to correct message printing
+    else:
+        context += ' '
```
This was just an annoying spacing issue with particular error messages

dotsdl commented:
Reviewing!

dotsdl commented on openforcefield/tests/test_toolkits.py:
```diff
@@ -1569,72 +1720,225 @@ def test_to_rdkit_losing_aromaticity_(self):
+
+    @pytest.mark.skipif(not RDKitToolkitWrapper.is_available() or not AmberToolsToolkitWrapper.is_available(),
+                        reason='RDKitToolkit and AmberToolsToolkit not available')
+    @pytest.mark.parametrize("partial_charge_method", ['am1bcc', 'am1-mulliken'])
```
Is it because `gasteiger` doesn't require conformers at all?

j-wags replied:
Yup. Added the same text to the docstring as above.

j-wags commented on openforcefield/typing/engines/smirnoff/parameters.py:
```diff
@@ -3083,12 +3113,17 @@ def create_force(self, system, topology, **kwargs):
-            temp_mol.generate_conformers(n_conformers=10)
-            temp_mol.compute_partial_charges(quantum_chemical_method=self._quantum_chemical_method,
-                                             partial_charge_method=self._partial_charge_method)
+            try:
+                # If the molecule wasn't assigned parameters from a manually-input charge_mol, calculate them here
+                temp_mol.generate_conformers(n_conformers=self.number_of_conformers)
+                temp_mol.assign_partial_charges(partial_charge_method=self.partial_charge_method)
```
Good call, and great attention to detail. I actually think this kwarg is pretty misleading, because it's not propagated to `find_smarts_matches` or any other toolkit-dependent functionality (including `compute_partial_bond_orders`!). But we already did it for `ToolkitAM1BCCHandler`, and we should continue the pattern, so I'll accept this patch. We should do what we can to expand this keyword out to all reasonable toolkit-dependent functionality in the future.

j-wags commented on openforcefield/utils/toolkits.py:
```diff
@@ -1756,7 +1922,50 @@ def compute_partial_charges_am1bcc(self, molecule):
             charge = charge * unit.elementary_charge
```
Relative to the other steps in parameterization, the cost of this step should be negligible. To me, the biggest factor here is readability/clarity.

j-wags commented on openforcefield/topology/molecule.py:
```diff
@@ -1519,36 +1537,55 @@ def __init__(self,
-                    result = toolkit_registry.call('from_object', other, allow_undefined_stereo=allow_undefined_stereo)
-                except NotImplementedError:
-                    pass
+                    result = toolkit_registry.call('from_object',
```
Oh, I thought about this so much, I didn't write down what the _successful_ logic means. D'oh!

Added 
```
# Each ToolkitWrapper may provide a from_object method, which turns some particular type(s)
# of object into OFFMols. For example, RDKitToolkitWrapper's from_object method will
# return an OFFMol if provided with an RDMol, or raise a ValueError if it is provided
# an OEMol (or anything else). This makes the assumption that any non-ValueError errors raised
# by the toolkit _really are_ bad and should be raised immediately, which may be a bad assumption.
```

j-wags commented on openforcefield/tests/test_toolkits.py:
```diff
@@ -657,77 +659,201 @@ def test_generate_conformers(self):
+        assert -1.e-5 < charge_sum.value_in_unit(unit.elementary_charge) < 1.e-5
+
+    @pytest.mark.skipif(not OpenEyeToolkitWrapper.is_available(), reason='OpenEye Toolkit not available')
+    @pytest.mark.parametrize("partial_charge_method", ['am1bcc', 'am1-mulliken'])
```
Yes -- Gasteiger isn't conformer-dependent, so this test (ensuring that charge methods generate different charges for different conformations) doesn't apply to it. Added
`Skip Gasteiger because it isn't conformer dependent.`
to the test docstring

j-wags commented on openforcefield/tests/test_forcefield.py:
```diff
@@ -23,11 +23,12 @@
 
-from openforcefield.utils.toolkits import OpenEyeToolkitWrapper, RDKitToolkitWrapper, AmberToolsToolkitWrapper, ToolkitRegistry
+from openforcefield.utils.toolkits import OpenEyeToolkitWrapper, RDKitToolkitWrapper, AmberToolsToolkitWrapper, \
+    ToolkitRegistry, ChargeMethodUnavailableError
```
Added in 49d4dc3

j-wags commented on openforcefield/tests/test_forcefield.py:
```diff
@@ -23,11 +23,12 @@
-from openforcefield.typing.engines.smirnoff import ForceField, ParameterHandler, IncompatibleParameterError, SMIRNOFFSpecError
-from openforcefield.typing.engines.smirnoff import XMLParameterIOHandler
+from openforcefield.typing.engines.smirnoff import ForceField, IncompatibleParameterError, SMIRNOFFSpecError, \
+    XMLParameterIOHandler, ParameterHandler
```
Updated in 49d4dc3

j-wags commented on openforcefield/typing/engines/smirnoff/parameters.py:
```diff
@@ -3013,30 +3023,30 @@ class ChargeIncrementType(ParameterType):
+        charge_increment = IndexedParameterAttribute(unit=unit.elementary_charge)
+
+        def __init__(self, **kwargs):
+            from openforcefield.typing.chemistry import ChemicalEnvironment
```
Removed in 49d4dc3

j-wags commented on openforcefield/utils/toolkits.py:
```diff
@@ -1756,7 +1922,50 @@ def compute_partial_charges_am1bcc(self, molecule):
             charges[index] = charge
 
-        return charges
+        if ((charges / unit.elementary_charge) == 0.).all():
```
Great call. Removed in 49d4dc3

j-wags commented on openforcefield/utils/toolkits.py:
```diff
@@ -3567,178 +3789,162 @@ def compute_partial_charges(self, molecule, charge_model=None):
+                ## TODO: How should we handle multiple conformers?
+                self._rdkit_toolkit_wrapper.to_file(
+                    mol_copy, 'molecule.sdf', file_format='sdf')
+                #os.system('ls')
```
Removed in 49d4dc3

j-wags commented on openforcefield/utils/toolkits.py:
```diff
@@ -3567,178 +3789,162 @@ def compute_partial_charges(self, molecule, charge_model=None):
-            The charge model to use. One of ['gas', 'mul', 'bcc']. If None, 'bcc' will be used.
-
+        partial_charge_method : str, optional, default=None
+            The charge model to use. One of ['gasteiger', 'am1-mulliken']. If None, 'am1-mulliken' will be used.
```
Whoops! Added in 49d4dc3

j-wags commented on openforcefield/topology/molecule.py:
```diff
@@ -2311,67 +2376,69 @@ def compute_partial_charges_am1bcc(self, toolkit_registry=GLOBAL_TOOLKIT_REGISTR
+        use_conformers : iterable of simtk.unit.Quantity-wrapped numpy arrays, each with shape (n_atoms, 3) and dimension of distance. Optional, default=None
+            Coordinates to use for partial charge calculation. If None, an appropriate number of conformers will be generated.
         toolkit_registry : openforcefield.utils.toolkits.ToolkitRegistry or openforcefield.utils.toolkits.ToolkitWrapper, optional, default=None
             :class:`ToolkitRegistry` or :class:`ToolkitWrapper` to use for SMILES-to-molecule conversion
```
Good catch. Fixed in 49d4dc3

j-wags commented on openforcefield/utils/toolkits.py:
```diff
@@ -1613,128 +1799,108 @@ def compute_partial_charges(self, molecule, quantum_chemical_method="AM1-BCC", p
+        if partial_charge_method == "am1bccnosymspt":
+            optimize = False
+            symmetrize = False
+            quacpac_status = oequacpac.OEAssignCharges(oemol, charge_method['oe_charge_method'](optimize, symmetrize))
```
Added 
```
# The OpenFF toolkit has always supported a version of AM1BCC with no geometry optimization
# or symmetry correction. So we include this keyword to provide a special configuration of quacpac
# if requested.
```

j-wags commented:
Worked on the inline comments today. I'll pick back up on your "overall" comments tomorrow!

j-wags commented:
> Should we add an explicit test for ToolkitWrapper._check_n_conformers? In fact, that method could be made into a staticmethod, since it really doesn't depend on self at all. A test of this should check that all the logic flow control works as expected.

Regarding the tests, good idea. I've added them

Regarding converting to staticmethod, I don't think it's a good long-term idea. We're going to convert ToolkitWrappers to have a cache of previously-performed operations, which would forbid us from using staticmethods and classmethods (since the only place a static or classmethod could store a cache is on the class itself, which is undesirable). For that reason, we're moving toward making all ToolkitWrapper methods into instance methods. 

> When are we planning to deprecate ToolkitAM1BCC?

Since it's in the already-published Parsley and smirnoff99Frosst FFs, never. 

> What do we expect to happen if the ChargeIncrementModel section in a SMIRNOFF spec has number_of_conformers=0? Does our machinery currently handle this case, as we'd expect for the gasteiger method?

I've updated the `formal_charges` example XML to have 0 confs defined (which it should have always been. D'oh!) to ensure that we can correctly handle cases where a FF defined `n_conformers=0`

> Remove the warning in the SMIRNOFF spec that ChargeIncrementModel is not yet implemented.

Good catch. Removed.

This should address everything you pointed out. I'll be leading the 0.7.0 release, so let's work through #572 first and I'll resolve any conflicts with this once it's merged.
